### PR TITLE
Heedls 408: Fix change password bug introduced in previous HEEDLS-408 work

### PR DESCRIPTION
**Pull request by SteveJacksonSoft**
### JIRA link
[_HEEDLS-408_](https://softwiretech.atlassian.net/browse/HEEDLS-408)

### Description
A type handler was added in the original HEEDLS-408 work which overrode how Dapper works with parameters of type `IEnumerable<int>`. This broke change-password.
I've replaced the type handler with a parse in the constructor of `Objective`.

I've also added a couple of tests and made the `ObjectiveBase` class abstract, since we aren't instantiating it and I don't expect we ever will.

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [ ] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Data.Tests/DataServices/PasswordDataServiceTests.cs [hunk 1]
```diff
@@ -156,5 +187,29 @@ public async Task Can_set_password_for_delegate_only()
             userDataService.GetAdminUserById(UserTestHelper.GetDefaultAdminUser().Id)?.Password.Should()
                 .NotBe(newPasswordHash);
         }
+
+        private void GivenDelegateUserIsInDatabase(DelegateUser user)
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data.Tests/DataServices/PasswordDataServiceTests.cs:191]: I should pull this out into a test helper.

File: DigitalLearningSolutions.Data.Tests/DataServices/PasswordDataServiceTests.cs [hunk 2]
```diff
@@ -95,6 +100,32 @@ public async Task Setting_password_by_email_sets_password_for_matching_candidate
...
+
+            // When
+            await passwordDataService.SetPasswordForUsersAsync(
+                new[] { existingDelegate.ToUserReference(), newDelegate.ToUserReference() },
+                newPasswordHash
+            );
+
+            // Then
+            userDataService.GetDelegateUserById(existingDelegate.Id)?.Password.Should()
+                .Be(PasswordHashNotYetInDb);
```
Comment (JBsoftwire) [inline DigitalLearningSolutions.Data.Tests/DataServices/PasswordDataServiceTests.cs:126]: Doesn't this also need an assert for the new delegate?

**Review by JBsoftwire — CHANGES_REQUESTED** [review]

**Review by JBsoftwire — APPROVED** [review]
Looks good

**Review by AlexJacksonDS — CHANGES_REQUESTED** [review]
Just a couple of minor things

File: DigitalLearningSolutions.Data/Models/Tracker/Objective.cs [hunk 3]
```diff
@@ -1,14 +1,28 @@
...
+        /// <summary>
+        /// Constructor for Dapper with signature matching object returned from get-objectives query.
+        /// </summary>
+        /// <param name="tutorialId"></param>
+        /// <param name="interactions">A comma-separated list of interactions, as stored in the db.</param>
+        /// <param name="possible"></param>
+        public Objective(int tutorialId, string? interactions, int possible) : base(tutorialId, possible)
+        {
+            Interactions = interactions?.Split(',', StringSplitOptions.RemoveEmptyEntries).Select(int.Parse).ToList() ??
+                           new List<int>();
```
[the remark this answers is not in the document]
Comment (kevwhitt-hee) [reply DigitalLearningSolutions.Data/Models/Tracker/Objective.cs:23]: Great news!

**Review by kevwhitt-hee — APPROVED** [review]
Looks good to me.

File: DigitalLearningSolutions.Data.Tests/DataServices/PasswordDataServiceTests.cs [hunk 4]
```diff
@@ -4,20 +4,24 @@
...
     using FluentAssertions;
+    using Microsoft.Data.SqlClient;
     using NUnit.Framework;
 
     public class PasswordDataServiceTests
     {
         private const string PasswordHashNotYetInDb = "I haven't used this password before!";
         private PasswordDataService passwordDataService = null!;
         private UserDataService userDataService = null!;
+        private SqlConnection? connection;
```
[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data.Tests/DataServices/PasswordDataServiceTests.cs:19]: @AlexJacksonDS What are we doing with the other things we're setting up?

[the remark this answers is not in the document]
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/DataServices/PasswordDataServiceTests.cs:19]: e.g. `private PasswordDataService passwordDataService = null!;`

**Review by AlexJacksonDS — APPROVED** [review]

**Pull request merged**